doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,260 @@
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Is the Surfaces on board with the proposed `CatalogItem` extensions?
+- Are there alternative proposals that we should consider?
+
+# Future work
+[future-work]: #future-work
+
+Note that `UsedIndexes` is currently computed inside `optimize_dataflow_index_imports`, which is  called towards the end of the “global” MIR optimization phase in `optimize_dataflow`.
+This will change with MaterializeInc/materialize#16598 and will most certainly cause conflicts if we don't coordinate the work between MaterializeInc/materialize#20652 (this doc) and MaterializeInc/materialize#16598.
+
+---
+
+Some more changes will be required if we decide to implement plan pinning. A naive approach would be to the LIR plan to `SerializedCatalogItem::V2`. A question remains what to do the other `ExplainInfo` fields. If we don’t serialize them, but instead try re-construct them during catalog re-hydration, we won’t get a faithful representation of the running plans in the `EXPLAIN <catalog item>` output.
+
+# Appendix A: `ExplainConfig` support
+
+The below code snippet summarizes which `ExplainConfig` options will be supported in the new `EXPLAIN <catalog item>` syntax.
+
+```rust
+pub struct ExplainConfig {
+    pub arity: bool,           // supported
+    pub join_impls: bool,      // supported
+    pub keys: bool,            // supported
+    pub linear_chains: bool,   // supported
+    pub non_negative: bool,    // supported
+    pub no_fast_path: bool,    // ignored (there is no fast path either way)
+    pub raw_plans: bool,       // supported
+    pub raw_syntax: bool,      // supported
+    pub subtree_size: bool,    // supported
+    pub timing: bool,          // ignored
+    pub types: bool,           // supported
+    pub filter_pushdown: bool, // unclear
+    pub cardinality: bool,     // ignored
+}
+```
+
+Note that we need need to be careful with the `filter_pushdown` handling. There are some options:
+
+1. Don't respect the `filter_pushdown` value. Treat it as `true` for catalog items that were created with MFP pushdown enabled, and as `false` otherwise.
+2. Don't respect the `filter_pushdown` value. Treat it as `false` for all catalog items.
+3. Continue respecting the `filter_pushdown` value. Treat it as "what if" condition that explains what would be pushed.
+
+My preferred behavior is option (1). Option (2) is a possibility if option (1) is too complicated. I have rejected option (3) because this is an instance of the behavior that we want to deprecate with #20652.
+
+In terms of implementation guidelines for option (1), as long as re-optimization continues to happen in the catalog re-hydration phase, it should be OK to just hard-code the value of the `filter_pushdown` flag based on the current persist MFP behavior (enabled or not).
```

Comment:
@bkirwi: what is the current way to encode 

> the current persist MFP behavior (enabled or not)

based on the current set of feature flag values?

Reply:
Stepping back a bit, the filter pushdown stuff has two parts:
1. The static-analysis bit: given a particular query, which of these filters can we analyze well enough to push down a filter for. This is a pure function of the query.
2. The dynamic bit: whether at runtime we fetch or ignore a particular part. This depends on lots of things: the current value of flags, the _historical_ value of flags, what the write patterns are in practice...

The current explain behaviour represents mostly 1, but also a little bit of 2 since it will behave differently depending on the live value of one flag. I think it would be fine to support this flag if it represented only 1, since it's just another way of representing data already in the plan. So I would propose to change the explain behaviour to eliminate the small dependency on 2.

I think the only thing that would need to change is to remove the check of the filtering-enabled flag at query parse time, when deciding whether/not to ignore the `filter_pushdown` explain option. If we did that, I believe your third option would be acceptable. Thoughts?